ledger/src/blockstore_metrics.rs

```diff
@@ -245,19 +249,30 @@ thread_local! {static PER_THREAD_ROCKS_PERF_CONTEXT: RefCell<PerfContext> = RefC
 /// and the PerfContext of the ubsequent RocksDB operation will be collected.
 pub(crate) fn maybe_enable_rocksdb_perf(
     sample_interval: usize,
-    perf_samples_counter: &AtomicUsize,
+    perf_status: &BlockstoreRocksDbPerfSamplingStatus,
 ) -> bool {
     if sample_interval == 0 {
         return false;
     }
 
-    if perf_samples_counter.fetch_add(1, Ordering::Relaxed) < sample_interval {
+    if perf_status.op_count.fetch_add(1, Ordering::Relaxed) < sample_interval {
         return false;
     }
     // Ideally, fetch_sub(*sample_interval) should be used to keep it
     // super precise.  However, since we do not use Mutex to protect the
     // above check and the below operation, we simply reset it to 0.
-    perf_samples_counter.store(0, Ordering::Relaxed);
+    perf_status.op_count.store(0, Ordering::Relaxed);
+
+    // Rate-limiting the perf sampling based on the minimum time interval.
+    let current_time_ms = timestamp();
+    if current_time_ms - perf_status.last_sample_time_ms.load(Ordering::Relaxed)
+        < ROCKS_PERF_METRIC_MIN_INTERVAL_MS
+    {
+        return false;
+    }
+    perf_status
+        .last_sample_time_ms
```

Comment:
instead of `last_sample_time_ms` you can store a timestamp for the next sample,
and use a single atomic `fetch_max` op to check and update:
https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU64.html#method.fetch_max

as it stands, this is doing two separate `load` and `store` which is not atomic.

Reply:
But if we do `fetch_max`, we will also update the counter even if we don't do the perf sample (and thus never do a perf sample unless we have two reads/writes that are one second apart, am I right at this part?)

Reply:
yeah, seems like `fetch_max` will not cut it.
This also needs to do [`compare_exchange_weak`](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU64.html#method.compare_exchange_weak) if we care about extra erroneous sampling/updates here.

Reply:
`compare_exchange_weak` only allows equality check, but here we need greater-than or less-than instead.  So we can't use `compare_exchange_weak` here either.

Since RocksDB's PerfContext is already thread-safe and we are okay with additional samples in rare cases, the current approach seems to be the best solution for performance at this moment.

Reply:
You can check for `>=` in the update logic. Something like:
```rust
let now = timestamp();
let old = self.last_sample_time_ms.load(Ordering::Relaxed);
if now.saturating_sub(old) < ROCKS_PERF_METRIC_MIN_DURATION_MS {
  return false;
}
loop {
    match self.last_sample_time_ms.compare_exchange_weak(
           old, now, Ordering::Relaxed, Ordering::Relaxed) {
        Ok(_) => break,
        Err(last_sample_time_ms) => {
          if last_sample_time_ms != old {
             // some other thread updated last_sample_time_ms
             return false;
          }
        }
    }
}
```

Reply:
Interesting.  Let me think more about it.